slither/detectors/operations/cache_array_length.py

```diff
@@ -0,0 +1,215 @@
+        array size could potentially change in that loop.
+        If it cannot, the loop condition is added to `non_optimal_array_len_usages`.
+        There may be some false negatives here - see docs for `_is_loop_referencing_array_length` for more information.
+        """
+        for node in nodes:
+            if node.type == NodeType.STARTLOOP:
+                if_node = node.sons[0]
+                if if_node.type != NodeType.IFLOOP:
+                    continue
+                if not isinstance(if_node.expression, BinaryOperation):
+                    continue
+                exp: BinaryOperation = if_node.expression
+                if not CacheArrayLength._is_identifier_member_access_comparison(exp):
+                    continue
+                array: StateVariable
+                if isinstance(exp.expression_right, MemberAccess):
+                    array = CacheArrayLength._extract_array_from_length_member_access(
+                        exp.expression_right
+                    )
+                else:  # isinstance(exp.expression_left, MemberAccess) == True
+                    array = CacheArrayLength._extract_array_from_length_member_access(
+                        exp.expression_left
+                    )
+                if array is None:
+                    continue
+
+                visited: Set[Node] = set()
+                if not CacheArrayLength._is_loop_referencing_array_length(
+                    if_node, visited, array, 1
+                ):
+                    non_optimal_array_len_usages.append(if_node.expression)
+
+    @staticmethod
+    def _get_non_optimal_array_len_usages_for_function(f: Function) -> List[SourceMapping]:
+        """
+        Finds non-optimal usages of array length in loop conditions in a given function.
+        """
+        non_optimal_array_len_usages: List[SourceMapping] = []
+        CacheArrayLength._handle_loops(f.nodes, non_optimal_array_len_usages)
+
+        return non_optimal_array_len_usages
+
+    @staticmethod
+    def _get_non_optimal_array_len_usages(functions: List[Function]) -> List[SourceMapping]:
+        """
+        Finds non-optimal usages of array length in loop conditions in given functions.
+        """
+        non_optimal_array_len_usages: List[SourceMapping] = []
+
+        for f in functions:
+            non_optimal_array_len_usages += (
+                CacheArrayLength._get_non_optimal_array_len_usages_for_function(f)
+            )
+
+        return non_optimal_array_len_usages
+
+    def _detect(self):
+        results = []
+
+        non_optimal_array_len_usages = CacheArrayLength._get_non_optimal_array_len_usages(
+            self.compilation_unit.functions
+        )
+        for usage in non_optimal_array_len_usages:
+            info = [
+                f"Loop condition at {usage.source_mapping} should use cached array length instead of referencing "
```

Comment:
```suggestion
                "Loop condition at", usage, "should use cached array length instead of referencing "
```
This will display more info context

Reply:
When I make that change, I'm getting `slither.exceptions.SlitherError: <class 'slither.core.expressions.binary_operation.BinaryOperation'> cannot be converted (no name, or canonical_name`. It seems that `BinaryOperation` isn't yet supported in `_convert_to_description`.

Reply:
https://github.com/crytic/slither/pull/1694#discussion_r1194131267